### PR TITLE
feat: add pop out mode

### PR DESCRIPTION
**Pull request by svnv**
Ref. https://github.com/Dintero/epics/issues/230
Rel. https://github.com/Dintero/checkout/pull/1107

Adds new functionality in the SDK where the checkout has both an embedded part but the actual payment flow is completed in a pop out. In order for this to work the button to launch the pop out is rendered client-side in the same runtime as the SDK. This is needed so that we can reliably launch and interact with the pop out from the SDK run-time. When the pop out is opened we also add a backdrop covering the entire screen for the current website. This should hinder the end user from changing the content of the session in the client while the user is completing the payment.

When the Checkout is opened in pop out mode:
- Payment button design is controlled by the embedded checkout
- Button and backdrop text is controlled by the embedded checkout  
- Apple pay should work with all compatible safari browsers as it is not from inside an iframe
- Walley Bank ID flow is no longer embedded (might solve some issues for Swedish end useres)
- The Klarna and Google Pay backdrop will cover our  entire payment pop out instead of just the embedded component (while our backdrop in turn covers the client page running the Dintero SDK)

### CODE REVIEW

File: test/dintero-checkout-web-sdk.test.ts [hunk 1]
```diff
@@ -51,33 +53,33 @@ const getHtmlBlobUrl = (
...
+//             "windowLocationAssign"
+//         );
+//         dintero.redirect({ sid: "<session_id>", language: "no" });
+//         sinon.assert.alwaysCalledWithExactly(
+//             windowLocationAssignStub,
+//             `https://checkout.dintero.com/v1/view/<session_id>?language=no&sdk=${pkg.version}`
+//         );
+//         windowLocationAssignStub.restore();
+//     });
+// });
```
Comment (andersem) [inline test/dintero-checkout-web-sdk.test.ts:82]: This doesn't work anymore?

**Review by andersem — APPROVED** [review]

File: src/url.ts [hunk 2]
```diff
@@ -17,19 +17,45 @@ export interface SessionUrlOptions {
...
+    return `${endpoint}/?sid=${sid}${params ? "&" + params : ""}`;
+};
+
+const padTralingSlash = (endpoint: string) => endpoint.endsWith('/') ? endpoint : `${endpoint}/`;
+
+export const getPopOutUrl = ({ sid, endpoint, language, shouldCallValidateSession }: SessionUrlOptions) => {
+    if (shouldCallValidateSession) {
+        return `${padTralingSlash(endpoint)}?loader=true`;
+    }
+    const params = new URLSearchParams();
```
Comment (andersem) [inline src/url.ts:52]: Use `URLSearchParams` in `getSessionUrl` as well for consistency?

**Review by andersem — APPROVED** [review]

File: src/popOut.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+const WIDTH = Math.min(480, window.screen.width);
+const HEIGHT = Math.min(840, window.screen.height);
+let popOutWindow: undefined | Window;
+
+const createPopOutWindow = (url: string, width: number, height: number) => {
+    // Opens a centered pop up window
+    const left = window.screenX + (window.outerWidth - width) / 2;
+    const top = window.screenY + (window.outerHeight - height) / 2;
+    const features = `width=${width},height=${height},left=${left},top=${top},location=no,menubar=no,toolbar=no,status=no`;
+    return window.open(url, 'dintero-checkout', features);
```
[the remark this answers is not in the document]
Comment (svnv) [reply src/popOut.ts:12]: Hmm, I will look into it. 

Some other options:
- revert to non-pop out embed when pop out fails (might not be a good idea with PCIDSS 4.0)
- show a link that redirects the user to the checkout

File: src/popOut.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if (popOutWindow && popOutWindow.closed) {
+            cleanUpClosed();
+        }
+    }
+
+    // Close pop out if current window is closed
+    window.addEventListener('beforeunload', closePopOut);
+
+    // Check if checkout is still open
+    intervalId = window.setInterval(checkIfPopupClosed, 50);
```
[the remark this answers is not in the document]
Comment (svnv) [reply src/popOut.ts:68]: The operation itself is not very CPU intensive, I'm not super worried about this. What would you suggest as a better default?

File: src/popOut.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (svnv) [reply src/popOut.ts:12]: We will revert to regular embed if pop out is not opened

File: src/popOut.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (svnv) [reply src/popOut.ts]: Decided on 200ms for now.

**Pull request merged**